app/controllers/ApplicationController.java

```diff
@@ -25,16 +30,66 @@ public ApplicationController(final Injector injector, final SetupController setu
         this.setupController = setupController;
     }
 
-    public F.Promise<Result> index() {
-        return setupController.handleOrFallback(() -> {
-            final HomeController homeController = injector.instanceOf(HomeController.class);
-            final ProjectContext projectContext = injector.instanceOf(ProjectContext.class);
-            final String defaultLanguage = projectContext.defaultLanguage().toLanguageTag();
-            return homeController.show(defaultLanguage);
-        });
-    }
-
     public Result untrail(final String path) {
         return movedPermanently("/" + path);
     }
+
+    public Result index() {
+        return setupController.handleOrFallback(() -> redirectToHomePage(defaultLanguage()));
+    }
+
+    public Result changeLanguage() {
+        final Form<LanguageFormData> languageForm = Form.form(LanguageFormData.class).bindFromRequest();
+        final String languageTag = languageForm.hasErrors() ? defaultLanguage() : languageForm.get().getLanguage();
+        return redirectToHomePage(languageTag);
+    }
+
+    public Result changeCountry(final String languageTag) {
+        final Form<CountryFormData> countryForm = Form.form(CountryFormData.class).bindFromRequest();
+        final String country = countryForm.hasErrors() ? defaultCountry() : countryForm.get().getCountry();
+        session(SESSION_COUNTRY, country);
+        Logger.debug("Changed country: " + session().toString());
+        return redirectToHomePage(languageTag);
+    }
+
+    private Result redirectToHomePage(final String languageTag) {
+        final ReverseRouter reverseRouter = injector.instanceOf(ReverseRouter.class);
+        return redirect(reverseRouter.home(languageTag));
+    }
+
+    private String defaultLanguage() {
+        final ProjectContext projectContext = injector.instanceOf(ProjectContext.class);
+        return projectContext.defaultLocale().toLanguageTag();
+    }
+
+    private String defaultCountry() {
+        final ProjectContext projectContext = injector.instanceOf(ProjectContext.class);
+        return projectContext.defaultCountry().getAlpha2();
+    }
+
+    public static class LanguageFormData extends Base {
+        @Constraints.Required
+        private String language;
+
+        public String getLanguage() {
+            return language;
+        }
+
+        public void setLanguage(final String language) {
+            this.language = language;
+        }
+    }
+
+    public static class CountryFormData extends Base {
+        @Constraints.Required
```

Comment:
It will be saved "1", but when Sunrise tries to use it, it will be converted into `CountryCode.UNDEFINED`, which will be discarded eventually and used the project's default country instead. I think it's not that important evaluating it here with regex or similar, when the evaluation will happen later anyway. The same applies to languages.